lab_5_scrapper/scrapper.py

```diff
@@ -120,46 +258,226 @@ def __init__(self, full_url: str, article_id: int, config: Config) -> None:
+        self._handle_crawler_data('w')
+
+    def load_crawler_data(self) -> None:
+        """
+        Loads start_url and collected urls
+        from a json file into crawler
+        """
+        self._handle_crawler_data('r')
+
+    def update_file_index(self) -> None:
+        """
+        Updates index of the last parsed article file
+        """
+        self._handle_crawler_data('w')
+
+    def delete_crawler_data(self) -> None:
+        """
+        Deletes the crawler_data.json file
+        """
+        if self.crawler_data_path.exists():
+            self.crawler_data_path.unlink()
+
+    def find_articles(self) -> None:
+        """
+        Recursive collecting and searching for links on the site
+        """
+        if self.num_visited_urls:
+            self.start_url = self.visited_urls[self.num_visited_urls - 1]
+        response = make_request(self.start_url, self._config)
+        links_bs = BeautifulSoup(response.content, 'lxml')
+        for link in links_bs.find_all('a'):
+            if self._extract_url(link):
+                url = self._extract_url(link)
+                if url and url not in self.urls \
+                        and len(self.urls) < self._config.get_num_articles():
+                    self.urls.append(url)
+            else:
+                href = link.get("href")
+                if not href \
+                        or href in self.visited_urls \
+                        or not href.startswith('https://chelny-izvest.ru/news/'):
+                    continue
+                self.visited_urls.append(href)
+        self._save_crawler_data()
+        while len(self.urls) < self._config.get_num_articles():
+            self.num_visited_urls += 1
+            self.find_articles()
 
 
 def main() -> None:
     """
     Entrypoint for scrapper module
     """
-    pass
+    config = Config(path_to_config=CRAWLER_CONFIG_PATH)
+    prepare_environment(ASSETS_PATH)
+    crawler = Crawler(config=config)
+    crawler.find_articles()
+    for ind, url in enumerate(crawler.urls, start=1):
+        parser = HTMLParser(full_url=url, article_id=ind, config=config)
+        article = parser.parse()
+        if isinstance(article, Article):
+            to_raw(article)
+            to_meta(article)
+
+
+def main_recursive() -> None:
+    """
+    Entrypoint for scrapper module
+    """
+    configuration = Config(path_to_config=CRAWLER_CONFIG_PATH)
+    prepare_environment(ASSETS_PATH, recursion=True)
+    recursive_crawler = CrawlerRecursive(config=configuration)
+    if not recursive_crawler.crawler_data_path.exists():
+        recursive_crawler.last_file_index = 1
+        shutil.rmtree(ASSETS_PATH)
+        ASSETS_PATH.mkdir(parents=True)
+    else:
+        recursive_crawler.load_crawler_data()
+    recursive_crawler.find_articles()
+    for index in range(recursive_crawler.last_file_index,
+                       len(recursive_crawler.urls) + 1):
+        recursive_crawler.last_file_index = index
```

Comment:
looks like the usage of indices is redundant. if all these indices are needed for filtering, then this should happen in the crawler. crawler results in trustworthy article links that we believe to be what we need and just pass later to the parser